[Bug][Item] Fix Reviver Seed and endure triggering off of indirect damage

schmidtc1 opened the pull request:
<!-- (Once you have read these comments, you are free to remove them) -->
<!-- Feel free to look at other PRs for examples -->
<!--
Make sure the title includes categorization (choose the one that best fits):
-       [Bug]: If the PR is primarily a bug fix
-      [Move]: If a move has new or changed functionality
-   [Ability]: If an ability has new or changed functionality
-      [Item]: For new or modified items
-   [Mystery]: For new or modified Mystery Encounters
-      [Test]: If the PR is primarily adding or modifying tests
-     [UI/UX]: If the PR is changing UI/UX elements
-     [Audio]: If the PR is adding or changing music/sfx
-    [Sprite]: If the PR is adding or changing sprites
-   [Balance]: If the PR is related to game balance
- [Challenge]: If the PR is adding or modifying challenges
-  [Refactor]: If the PR is primarily rewriting existing code
-      [Docs]: If the PR is just adding or modifying documentation (such as tsdocs/code comments)
-    [GitHub]: For changes to GitHub workflows/templates/etc
-      [Misc]: If no other category fits the PR
-->
<!--
Make sure that this PR is not overlapping with someone else's work
Please try to keep the PR self-contained (and small)
-->
TODO:

- [x] Revise code to proc Reviver Seed on confusion, ignore all other indirect damage
- [x] Make OHKO moves not bypass Endure or Reviver Seed
- [x] Test manually
  - [x] Should activate RSeed
    - [x] "Normal" damaging moves
    - [x] OHKO moves
    - [x] Fixed damage
      - [x] Seismic toss
      - [x] Super fang
    - [x] Final Gambit
      - [x] Damage portion (self-faint portion should not activate RSeed)
    - [x] Confusion self-hit
    - [x] Delayed attacks (Future Sight)
    - [x] Counter moves
  - [x] Should not active RSeed
    - [x] Status damage (burn and poison)
    - [x] Hail/Sand chip
    - [x] Salt Cure, Curse, Leech Seed chip
      - [x] Salt Cure
      - [x] Curse
      - [x] Leech Seed
    - [x] Partial trapping chip (Magma Storm, Infestation, Whirlpool chip)
    - [x] Recoil damage
    - [x] Weather damage
    - [x] Deduction damage (Ghost Curse)
    - [x] Self-KOing (Explosion, Memento, Final Gambit)
    - [x] Damage from Liquid Ooze/Innards Out
      - [x] Fix Liquid Ooze 
      - [x] Innards Out
    - [x] Misc cases
      - [x] Fire Pledge
      - [x] Bad Dreams
- [x] Fix Endure from blocking indirect damage
- [x] Write/adjust unit tests
  - [x] Reviver Seed tests
  - [x] Endure tests


## What are the changes the user will see?
<!-- Summarize what are the changes from a user perspective on the application -->
Reviver Seed's description states it triggers off of direct damage, but it triggers from some forms of indirect damage.
Reviver Seed should no longer trigger off of indirect damage (weather, Iron Barbs ability, status effect, etc.)
This PR will attempt to cover these cases mentioned on Discord:
https://discord.com/channels/1125469663833370665/1176874654015684739/1332821527787208777


## Why am I making these changes?
<!--
Explain why you decided to introduce these changes
Does it come from an issue or another PR? Please link it
Explain why you believe this can enhance user experience
-->
Bug fix/matching implementation to description
<!--
If there are existing GitHub issues related to the PR that would be fixed,
you can add "Fixes #[issue number]" (ie: "Fixes #1234") to link an issue to your PR
so that it will automatically be closed when the PR is merged.
-->
Fixes #5114
Fixes #5115

## What are the changes from a developer perspective?
<!--
Explicitly state what are the changes introduced by the PR
You can make use of a comparison between what was the state before and after your PR changes
Ex: What files have been changed? What classes/functions/variables/etc have been added or changed?
-->
Created a `damageSources` array under `turnData` which tracks `DamageResult`. This will store the latest `DamageResult` sent to the `damageAndUpdate` function in `turnData.damageSources[0]`. If no `DamageResult` is supplied, `HitResult.OTHER` is used which will mark it as an indirect damage.
`PokemonInstantReviveModifier` now checks for `HitResult.OTHER` in its apply function. If the most recent damage taken is `HitResult.OTHER`, the apply function returns early and has no effect.
Liquid Ooze now uses `HitResult.OTHER`, instead of `HitResult.HEAL`. 

## Screenshots/Videos
<!--
If your changes are changing anything on the user experience, please provide visual proofs of it
Please take screenshots/videos before and after your changes, to show what is brought by this PR
-->
<details><summary>Status damage</summary>

https://github.com/user-attachments/assets/a9f792d0-d063-4d7e-bd62-e4d897c83b7b
</details>

<details><summary>Weather</summary>

https://github.com/user-attachments/assets/40f9be67-43a3-4205-a1e6-015b21b0eac5
</details>

<details><summary>OHKO</summary>

https://github.com/user-attachments/assets/7bc40b3d-f9fc-46bd-9eef-b839c366efdc
</details>

<details><summary>Endure</summary>

https://github.com/user-attachments/assets/b12d6af1-31a2-4c64-8060-2275a0cc4e56
</details>

## How to test the changes?
<!--
How can a reviewer test your changes once they check out on your branch?
Did you make use of the `src/overrides.ts` file?
Did you introduce any automated tests?
Do the reviewers need to do something special in order to test your changes?
-->

## Checklist
- [x] **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I tested the changes manually?
- [x] Are all unit tests still passing? (`npm run test`)
  - [x] Have I created new automated tests (`npm run create-test`) or updated existing tests related to the PR's changes?
- [x] Have I provided screenshots/videos of the changes (if applicable)?
  - [n/a] Have I made sure that any UI change works for both UI themes (default and legacy)?

Are there any localization additions or changes? If so:
- [n/a] Has a locales PR been created on the [locales](https://github.com/pagefaultgames/pokerogue-locales) repo?
  - [n/a] If so, please leave a link to it here: 
- [n/a] Has the translation team been contacted for proofreading/translation?

Wlowscha commented on src/field/pokemon.ts:
```diff
@@ -5356,10 +5366,11 @@ export enum HitResult {
   OTHER,
-  IMMUNE
+  IMMUNE,
+  SELF
```
From what I see the SELF HitResult is only used for confusion; maybe it would be appropriate to rename it to CONFUSION to avoid mixing it up with other types of self-damage, e.g. from Explosion or Metal Burst

schmidtc1 replied:
Updated to CONFUSION.

Wlowscha commented:
I have gone through the PR and looked at the relevant code, and I don't quite understand why it is necessary to introduce `this.turnData.damageSources`.
This is checked twice:
1) In `Pokemon.damage()`, where `!preventEndure` is also checked in the same condition,
2) In `PokemonInstantReviveModifier.apply()`, which is only called once in the code from `FaintPhase` and is locked behind an `if (!this.preventEndure)` condition.

There are only two `new FaintPhase`, one in `Pokemon.damage()` (which already looks at a `preventEndure` variable) and one in `Pokemon.apply()` (which is being changed in this PR to always set `preventEndure` to `false`. Therefore, it would seem that passing the correct `preventEndure` to `Pokemon.damage()` should be enough.

`Pokemon.damage()` itself is only called in three places:
1) `Pokemon.damageAndUpdate()`
2) In the code for Pain Split
3) In the status damage at the end of turn (poison, toxic, burn)

For (2) and (3) it could be argued that the `Pokemon.damage()` calls should be replaced by `Pokemon.damageAndUpdate()` anyway, but in any case they are easy to manage. Pain Split doesn't really matter since it never kills, and the status damage is already passing `true` as `preventEndure`. 
`Pokemon.damageAndUpdate()` is taking both a `HitResult` parameter and a `preventEndure`, so it should be possible to check here if the result is `HitResult.OTHER`, and, in that case, pass on `preventEndure = true` if it wasn't already.
(In fact, it might be preferable to remove `preventEndure` altogether and only look at the `HitResult`, since this PR is going in the direction to ensure that all sources which would `preventEndure` are marked as `HitResult.OTHER`).


I'm saying all this mainly to see if a simpler solution can be found, as it seems that introducing the `turnData.damageSources` was the very first action taken. Are there any strange edge cases that necessarily require the `turnData.damageSources` in order to be handled? For example, is it possible that `Pokemon.apply()` is called but somehow the mon dies from indirect damage before going to the `FaintPhase`?

schmidtc1 commented:
> I'm saying all this mainly to see if a simpler solution can be found, as it seems that introducing the `turnData.damageSources` was the very first action taken. Are there any strange edge cases that necessarily require the `turnData.damageSources` in order to be handled? For example, is it possible that `Pokemon.apply()` is called but somehow the mon dies from indirect damage before going to the `FaintPhase`?

It was the first thing that came to mind for me. I figured it'd be easier to add the `HitResult` in the `damageAndUpdate` function rather than going to each individual call of `damageAndUpdate`. 

I don't think there are any edge cases requiring `turnData.damageSources`.

I think to prevent any significant overhaul to this PR, I would prefer to remove the `preventEndure` boolean from `damageAndUpdate` and `FaintPhase`. I don't like that we tie Reviver Seed's logic with `preventEndure`. Intuitively, I see a variable named `preventEndure` only applying to Endure. 

If you prefer to keep `preventEndure`, I'll work on that instead.

Wlowscha commented:
> > I'm saying all this mainly to see if a simpler solution can be found, as it seems that introducing the `turnData.damageSources` was the very first action taken. Are there any strange edge cases that necessarily require the `turnData.damageSources` in order to be handled? For example, is it possible that `Pokemon.apply()` is called but somehow the mon dies from indirect damage before going to the `FaintPhase`?
> 
> It was the first thing that came to mind for me. I figured it'd be easier to add the `HitResult` in the `damageAndUpdate` function rather than going to each individual call of `damageAndUpdate`.
> 
> I don't think there are any edge cases requiring `turnData.damageSources`.
> 
> I think to prevent any significant overhaul to this PR, I would prefer to remove the `preventEndure` boolean from `damageAndUpdate` and `FaintPhase`. I don't like that we tie Reviver Seed's logic with `preventEndure`. Intuitively, I see a variable named `preventEndure` only applying to Endure.
> 
> If you prefer to keep `preventEndure`, I'll work on that instead.

Note that if you actually wanted to remove `preventEndure`, you would then have to go through all the `damageAndUpdate` calls to make sure that argument is not passed anymore :)
But this was actually part of my suggestion: with your changes to various things like weather and liquid ooze, you are ensuring that all calls which would need to pass `preventEndure = true` to `damageAndUpdate` are also passing `HitType.OTHER`. Since "other === indirect damage === prevents endure", the `preventEndure` argument can be discarded from `damageAndUpdate`.

However, you still need to pass `preventEndure` to `damage`, which then passes it on to `FaintPhase`. If the issue is with the naming, you can rename it to `activatesReviverSeed` when passing it to `FaintPhase`, or even rename those instances to `isIndirectDamage` for example.

It seems to me that we should avoid adding a new attribute to the Pokemon class to track information which is anyway passed directly from `damageAndUpdate` to `damage` to `FaintPhase`, if that attribute serves no other purpose. It's another story if tracking that information can become somehow useful elsewhere, but at the moment that doesn't seem the case. I understand not wanting to make big changes to the PR though, I'll leave it to the others to weigh in :)

schmidtc1 commented:
Refactored code based on suggestions by @Wlowscha 
Removed damageSources array in favor of sending an isIndirectDamage boolean to faintPhase and damage whenever HitResult.OTHER is encountered or no HitResult is supplied. Removed the preventEndure variable as a parameter in damageAndUpdate, it now only appears in FaintPhase.

SirzBenjie reviewed (dismissed):
With @Wlowscha's suggestions integrated, this PR is very clean and lean. I like it.
Also you added good tests for it.

DayKev commented on src/field/pokemon.ts:
```diff
@@ -3163,13 +3161,21 @@ export default abstract class Pokemon extends Phaser.GameObjects.Container {
    * @returns integer of damage done
    */
-  damageAndUpdate(damage: number, result?: DamageResult, critical: boolean = false, ignoreSegments: boolean = false, preventEndure: boolean = false, ignoreFaintPhase: boolean = false, source?: Pokemon): number {
+  damageAndUpdate(damage: number, result?: DamageResult, critical: boolean = false, ignoreSegments: boolean = false, ignoreFaintPhase: boolean = false, source?: Pokemon): number {
```
It's probably a good idea to change the optional parameters to be an object, ie:
```suggestion
  damageAndUpdate(damage: number,
    {
      result = HitResult.EFFECTIVE,
      critical = false,
      ignoreSegments = false,
      ignoreFaintPhase = false,
    }:
    {
      result?: DamageResult,
      critical?: boolean,
      ignoreSegments?: boolean,
      ignoreFaintPhase?: boolean,
      source?: Pokemon,
     }
  ): number {
```

schmidtc1 replied:
Refactored in [855d795](https://github.com/pagefaultgames/pokerogue/pull/5182/commits/855d795f2d4774423ff6e19808c7d4ead0e11a97)

DayKev reviewed (commented):
Is `HitResult.INDIRECT_KO` only used for the damage number display and sfx?

schmidtc1 commented:
> Is `HitResult.INDIRECT_KO` only used for the damage number display and sfx?

Yes, this was discussed on discord. Perish song and destiny bond are unique KO moves that will not trigger reviver seed, but still have the KO sfx, so they were given the `INDIRECT_KO` hit result

schmidtc1 commented on src/field/pokemon.ts:
```diff
@@ -3163,13 +3161,21 @@ export default abstract class Pokemon extends Phaser.GameObjects.Container {
-  damageAndUpdate(damage: number, result?: DamageResult, critical: boolean = false, ignoreSegments: boolean = false, preventEndure: boolean = false, ignoreFaintPhase: boolean = false, source?: Pokemon): number {
+  damageAndUpdate(damage: number, result?: DamageResult, critical: boolean = false, ignoreSegments: boolean = false, ignoreFaintPhase: boolean = false, source?: Pokemon): number {
+    let isIndirectDamage: boolean = true;
+    if (result !== undefined) {
```
Fixed in [b5bf976](https://github.com/pagefaultgames/pokerogue/pull/5182/commits/b5bf9766aab5bbebaaffe6461911cc8e538b568b)

DayKev commented on src/field/pokemon.ts:
```diff
@@ -3175,13 +3173,21 @@ export default abstract class Pokemon extends Phaser.GameObjects.Container {
    * @returns integer of damage done
    */
-  damageAndUpdate(damage: number, result?: DamageResult, critical: boolean = false, ignoreSegments: boolean = false, preventEndure: boolean = false, ignoreFaintPhase: boolean = false, source?: Pokemon): number {
+  damageAndUpdate(damage: number,
```
You should rename `critical` to `isCritical`.

Also, don't forget to update the tsdocs for this function with the changes you've made.

schmidtc1 commented on src/data/battler-tags.ts:
```diff
@@ -723,7 +723,7 @@ export class DestinyBondTag extends BattlerTag {
       })
     );
-    pokemon.damageAndUpdate(pokemon.hp, HitResult.ONE_HIT_KO, false, false, true);
+    pokemon.damageAndUpdate(pokemon.hp, { result: HitResult.INDIRECT_KO, ignoreSegments: true });
```
- You told me on [discord](https://discord.com/channels/1125469663833370665/1176874654015684739/1342284309117145170) that Reviver Seed will trigger in any case that endure will trigger. Endure would not trigger off of Destiny Bond or Perish Song, since Endure can only trigger off of [attacks](https://bulbapedia.bulbagarden.net/wiki/Endure_(move)#:~:text=Endure%20allows%20the%20user%20to%20survive%20any%20attack%20that%20would%20cause%20it%20to%20faint). Thus, Reviver Seed would not trigger off of Destiny Bond or Perish Song.
- I have been told to make [OHKO moves trigger Reviver Seed](https://discord.com/channels/1125469663833370665/1176874654015684739/1332821527787208777). 

Based on these two requirements, I created `HitResult.INDIRECT_KO`, since actual OHKO moves would function differently than Destiny Bond and Perish Song.

If you still want Destiny Bond and Perish Song to trigger Reviver Seed, then I will change it.

DayKev replied:
Oh I missed one use in `Pokemon.getAttackDamage()`, I thought you replaced them all. I guess it's fine then, not sure there's a clean way to do it... sigh, thanks gamefreak.

schmidtc1 commented:
@DayKev's suggestions resolved in [3dfeabe](https://github.com/pagefaultgames/pokerogue/pull/5182/commits/3dfeabe46a9aba1914fece40fa3d1e1284624955) and [de8efba](https://github.com/pagefaultgames/pokerogue/pull/5182/commits/de8efba9fef430506ba4178d70663f8112ba9520)

damocleas commented:
now has a conflict, assuming related to #5495
